src/main/scala/com/mozilla/telemetry/pings/FocusEvent.scala

```diff
@@ -24,32 +32,66 @@ case class FocusEventPing(
     osversion: String,
     meta: Meta) {
 
-  def getEvents: String = {
-    val eventsList = events.map{ e =>
-      ("device_id" -> getHashedClientId) ~
-      ("event_type" -> e.getType) ~
-      ("time" -> (e.timestamp + created)) ~
-      ("event_properties" ->
-        e.extra.getOrElse(Map.empty).foldLeft(JObject())(_ ~ _) ~
-        ("object" -> e.`object`) ~
-        ("value" -> e.value)) ~
-      ("app_version" -> meta.appVersion) ~
-      ("os_name" -> os) ~
-      ("os_version" -> osversion) ~
-      ("country" -> meta.geoCountry) ~
-      ("city" -> meta.geoCity)
-    }
+  def getEvents(config: Config): String = {
+    implicit val formats = DefaultFormats
+
+    val factory = JsonSchemaFactory.byDefault
+    val schemas = config.events.map(e => factory.getJsonSchema(asJsonNode(e.schema)))
+
+    val clientid = getHashedClientId
+    val sessionId = getSessionId
+
+    val eventsList = events.map{ e => e -> asJsonNode(Extraction.decompose(e)): (Event, JsonNode) }
+        .map{ case(e, es) => // for each event, try each schema
+          e -> schemas.map( ts => ts.validateUnchecked(es).isSuccess )
+            .zip(config.events)
+            .filter(_._1)
+      }
+      .filter{ case(e, es) => !es.isEmpty } // only keep those with a match
+      .map{ case(e, es) => e -> es.head._2 } // take the first match (head._1 is the bool)
+      .map{ case(e, es) =>
+        ("device_id" -> clientId) ~
+        ("session_id" -> sessionId) ~
+        ("insert_id" -> (clientId + sessionId.toString + e.getId)) ~
+        ("event_type" -> getFullEventName(config.eventGroupName, es.name)) ~
+        ("time" -> (e.timestamp + created)) ~
+        ("event_properties" -> e.getProperties(es.amplitudeProperties)) ~
+        ("app_version" -> meta.appVersion) ~
+        ("os_name" -> os) ~
+        ("os_version" -> osversion) ~
+        ("country" -> meta.geoCountry) ~
+        ("city" -> meta.geoCity)
+      }
 
     compact(render(eventsList))
   }
 
+  def getFullEventName(groupName: String, eventName: String): String = groupName + " - " + eventName
+
+  def getSessionId: Long = events.map(_.timestamp).max
+
   def getHashedClientId: String = {
     val messageDigest = MessageDigest.getInstance("SHA-256")
     new String(messageDigest.digest(clientId.getBytes()))
   }
 
-  def includeClient(sample: Double): Boolean = {
-    meta.sampleId.getOrElse(sample * 100) < (sample * 100)
+  def includePing(sample: Double, config: Config): Boolean = {
+    val keepClient = meta.sampleId.getOrElse(sample * 100) < (sample * 100)
+
+    if(!keepClient){
+      // for maybe a slight perf increase
+      return false // scalastyle:ignore
```

Comment:
Since you're doing an early return due to potential perf improvements, it might make a difference to split this back out and move the sample check earlier in `parsePing` before iterating through the top-level filters

Reply:
I'll probably end up removing this anyways -- sounds like Amplitude recently enabled client sampling on their end, which means the numbers that show in the dashboards are more correct. For now, I'm going to keep it in just for testing purposes.